### PR TITLE
feat: Implement `instruction_addr_adjustment`

### PR DESCRIPTION
**Pull request by Swatinem**
This attribute is specified in [RFC 43](https://github.com/getsentry/rfcs/pull/43).

Am I doing something wrong? I’m shocked by the amount of boilerplate I have to cargo-cult just to get an enum into the protocol.

### CODE REVIEW

File: relay-general/src/protocol/stacktrace.rs [hunk 1]
```diff
@@ -350,6 +362,123 @@ pub struct RawStacktrace {
...
+/// stack frames, `"none"` should be used here.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+#[cfg_attr(feature = "jsonschema", derive(schemars::JsonSchema))]
+#[cfg_attr(feature = "jsonschema", schemars(rename_all = "snake_case"))]
+pub enum InstructionAddrAdjustment {
+    /// The default. Applies a heuristic based on other event / exception attributes.
+    Auto,
+
+    /// All but the first frame needs to be adjusted. The first frame is not a *return address*,
```
Comment (loewenheim) [inline relay-general/src/protocol/stacktrace.rs]:
```suggestion
    /// All but the first frame needs to be adjusted. The first frame's address is not a *return address*,
```

File: relay-general/src/protocol/stacktrace.rs [hunk 2]
```diff
@@ -350,6 +362,123 @@ pub struct RawStacktrace {
...
+    /// but points directly to the faulty instruction.
+    AllButFirst,
+
+    /// All frames should be adjusted, for example because the stack walking implementation truncated
+    /// frames from the top of the stack, and all remaining frames are *return addresses*.
```
Comment (loewenheim) [inline relay-general/src/protocol/stacktrace.rs]:
```suggestion
    /// frames from the top of the stack, and all remaining frames' addresses are *return addresses*.
```

File: relay-general/src/protocol/stacktrace.rs [hunk 3]
```diff
@@ -330,6 +337,11 @@ pub struct RawStacktrace {
     /// values of the thread, thus mapping to the last frame in the list.
     pub registers: Annotated<Object<RegVal>>,
 
+    /// Optional. A flat that indicates if, and how, `instruction_addr` values need to be adjusted
```
Comment (loewenheim) [inline relay-general/src/protocol/stacktrace.rs]:
```suggestion
    /// Optional. A flag that indicates if, and how, `instruction_addr` values need to be adjusted
```

**Review by loewenheim — APPROVED** [review]

File: relay-general/src/protocol/stacktrace.rs [hunk 4]
```diff
@@ -350,6 +362,123 @@ pub struct RawStacktrace {
...
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = match self {
+            InstructionAddrAdjustment::Auto => "auto",
+            InstructionAddrAdjustment::AllButFirst => "all_but_first",
+            InstructionAddrAdjustment::All => "all",
+            InstructionAddrAdjustment::None => "none",
+        };
+        f.write_str(s)
+    }
+}
```
Comment (jjbayer) [inline relay-general/src/protocol/stacktrace.rs]: `derive_fromstr_and_display` might be useful here, see for example https://github.com/getsentry/relay/blob/0df83a57639e00fa89a51c8e05123d7bba7c497d/relay-general/src/protocol/security_report.rs#L522

File: relay-general/src/protocol/stacktrace.rs [hunk 5]
```diff
@@ -350,6 +362,123 @@ pub struct RawStacktrace {
...
+
+impl Empty for InstructionAddrAdjustment {
+    #[inline]
+    fn is_empty(&self) -> bool {
+        matches!(self, Self::Auto)
+    }
+}
+
+impl FromValue for InstructionAddrAdjustment {
```
Comment (jjbayer) [inline relay-general/src/protocol/stacktrace.rs:441]: Have you tried deriving `FromValue`?

File: relay-general/src/protocol/stacktrace.rs [hunk 6]
```diff
@@ -350,6 +362,123 @@ pub struct RawStacktrace {
...
+    /// but points directly to the faulty instruction.
+    AllButFirst,
+
+    /// All frames should be adjusted, for example because the stack walking implementation truncated
+    /// frames from the top of the stack, and all remaining frames' addresses are *return addresses*.
+    All,
+
+    /// The stack walking implementation already provides correct addresses and no adjustment should
+    /// be performed when symbolicating.
+    None,
```
Comment (jjbayer) [inline relay-general/src/protocol/stacktrace.rs:392]: How likely is it that we will add new variants to this in the future? If there's a chance, we should make this enum forward-compatible by adding a fallback `Unknown(String)` variant here, such that outdated external Relays can still forward this data to internal Relays. See for example https://github.com/getsentry/relay/blob/0df83a57639e00fa89a51c8e05123d7bba7c497d/relay-general/src/protocol/transaction.rs#L32-L34

If we can live with outdated Relays not forwarding this information, we can also make deserialization fallback to `Auto`.

File: relay-general/src/protocol/stacktrace.rs (hunk 5 above)
Comment (loewenheim) [reply relay-general/src/protocol/stacktrace.rs:441]: Deriving `FromValue` and `IntoValue` unfortunately doesn't work. I don't understand the problem in the former case, but the latter seems to expect all enum variants to be tuple-like.

File: relay-general/src/protocol/stacktrace.rs (hunk 6 above)
Comment (loewenheim) [reply relay-general/src/protocol/stacktrace.rs:392]: I added an `Unknown` variant and made parsing infallible, is that what you had in mind?

Comment (jjbayer) [reply relay-general/src/protocol/stacktrace.rs:392]: Yes, thanks!

**Review by jjbayer — APPROVED** [review]

**Pull request merged**